File: meditor/src/osm/parser.ts
import type { Network, TrafficNode, TrafficLink, TransportRoute, Building, BuildingType, ProjectedCoords } from "../types";
import type { OSMElement, OSMNode, OSMWay, OSMRelation } from "../types/osm";
import {
  BUILDING_TAG_MAPPINGS,
  ID_PREFIXES,
  OSM_TAG_VALUES,
  GEOMETRY_VALIDATION,
} from "../constants";
import { wgs84ToProjected } from "../utils/coordinates";

function indexNodes(elements: OSMElement[]): Map<number, OSMNode> {
  const osmNodes = new Map<number, OSMNode>();
  for (const el of elements) {
    if (el.type === OSM_TAG_VALUES.TYPE_NODE) {
      osmNodes.set(el.id, el as OSMNode);
    }
  }
  return osmNodes;
}

function indexWays(elements: OSMElement[]): Map<number, OSMWay> {
  const osmWays = new Map<number, OSMWay>();
  for (const el of elements) {
    if (el.type === OSM_TAG_VALUES.TYPE_WAY) {
      osmWays.set(el.id, el as OSMWay);
    }
  }
  return osmWays;
}

function countNodeUsage(elements: OSMElement[]): Map<number, number> {
  const nodeUsage = new Map<number, number>();
  for (const el of elements) {
    if (el.type === OSM_TAG_VALUES.TYPE_WAY && el.tags?.highway) {
      const way = el as OSMWay;
      for (const nodeId of way.nodes) {
        nodeUsage.set(nodeId, (nodeUsage.get(nodeId) || 0) + 1);
      }
    }
  }
  return nodeUsage;
}

function buildGeometry(
  wayNodes: number[],
  osmNodes: Map<number, OSMNode>,
  targetCRS: string
): ProjectedCoords[] {
  const geometry: ProjectedCoords[] = [];
  for (const nodeId of wayNodes) {
    const osmNode = osmNodes.get(nodeId);
    if (osmNode) {
      // Transform WGS84 → projected CRS at data entry
      const [x, y] = wgs84ToProjected(osmNode.lat, osmNode.lon, targetCRS);
      geometry.push([x, y]);
    }
  }
  return geometry;
}

function createLink(
  way: { id: number; nodes: number[]; tags: Record<string, string> },
  geometry: ProjectedCoords[]
): TrafficLink {
  const fromNodeId = `${ID_PREFIXES.NODE}${way.nodes[0]}`;
  const toNodeId = `${ID_PREFIXES.NODE}${way.nodes[way.nodes.length - 1]}`;

  return {
    id: `${ID_PREFIXES.LINK}${way.id}`,
    osmId: way.id,
    from: fromNodeId,
    to: toNodeId,
    geometry,
    tags: {
      highway: way.tags.highway,
      lanes: way.tags.lanes ? parseInt(way.tags.lanes) : undefined,
      maxspeed: way.tags.maxspeed ? parseInt(way.tags.maxspeed) : undefined,
      oneway: way.tags.oneway === OSM_TAG_VALUES.ONEWAY_YES,
      name: way.tags.name,
    },
  };
}

function createNode(
  osmId: number,
  osmNode: OSMNode,
  connectionCount: number,
  targetCRS: string
): TrafficNode {
  const [x, y] = wgs84ToProjected(osmNode.lat, osmNode.lon, targetCRS);
  return {
    id: `${ID_PREFIXES.NODE}${osmId}`,
    osmId,
    position: [x, y],
    connectionCount,
  };
}

function createTransportRoute(
  relationId: number,
  wayId: number,
  geometry: ProjectedCoords[],
  tags: Record<string, string>
): TransportRoute {
  return {
    id: `${ID_PREFIXES.TRANSPORT}${relationId}${ID_PREFIXES.TRANSPORT_WAY}${wayId}`,
    osmId: relationId,
    wayId,
    geometry,
    tags: {
      route: tags.route,
      ref: tags.ref,
      name: tags.name,
      network: tags.network,
      operator: tags.operator,
      colour: tags.colour,
    },
  };
}

function determineBuildingType(tags: Record<string, string>): BuildingType | null {
  for (const mapping of BUILDING_TAG_MAPPINGS) {
    if (tags[mapping.tag] === mapping.value) {
      return mapping.type as BuildingType;
    }
  }
  return null;
}

function createBuilding(
  osmId: number,
  position: ProjectedCoords,
  tags: Record<string, string>,
  geometry?: ProjectedCoords[]
): Building | null {
  const type = determineBuildingType(tags);
  if (!type) return null;

  return {
    id: `${ID_PREFIXES.BUILDING}${osmId}`,
    osmId,
    position,
    geometry,
    type,
    tags: {
      name: tags.name,
      building: tags.building,
      shop: tags.shop,
      amenity: tags.amenity,
    },
  };
}

export function parseOSMResponse(elements: OSMElement[], targetCRS: string): Network {
  const nodes = new Map<string, TrafficNode>();
  const links = new Map<string, TrafficLink>();
  const transportRoutes = new Map<string, TransportRoute>();
  const buildings = new Map<string, Building>();
  const osmNodes = indexNodes(elements);
  const osmWays = indexWays(elements);
  const nodeUsage = countNodeUsage(elements);

  for (const el of elements) {
    if (el.type === OSM_TAG_VALUES.TYPE_WAY && el.tags?.highway) {
      const way = el as OSMWay;
      const geometry = buildGeometry(way.nodes, osmNodes, targetCRS);
      if (geometry.length < GEOMETRY_VALIDATION.MIN_LINK_POINTS) continue;

      const link = createLink(
        { id: way.id, nodes: way.nodes, tags: way.tags || {} },
        geometry
      );
      links.set(link.id, link);

      const endpoints = [way.nodes[0], way.nodes[way.nodes.length - 1]];
      for (const osmId of endpoints) {
        const nodeId = `${ID_PREFIXES.NODE}${osmId}`;
        if (nodes.has(nodeId)) continue;

        const osmNode = osmNodes.get(osmId);
        if (osmNode) {
          nodes.set(
            nodeId,
            createNode(osmId, osmNode, nodeUsage.get(osmId) || 1, targetCRS)
          );
        }
      }
    } else if (el.type === OSM_TAG_VALUES.TYPE_RELATION && el.tags?.route) {
      const relation = el as OSMRelation;
      for (const member of relation.members) {
        if (member.type !== OSM_TAG_VALUES.TYPE_WAY) continue;

        const way = osmWays.get(member.ref);
        if (!way || way.nodes.length < GEOMETRY_VALIDATION.MIN_ROUTE_POINTS) continue;

        const geometry = buildGeometry(way.nodes, osmNodes, targetCRS);
        if (geometry.length < GEOMETRY_VALIDATION.MIN_ROUTE_POINTS) continue;

        const route = createTransportRoute(
          relation.id,
          member.ref,
          geometry,
          relation.tags || {}
        );
        transportRoutes.set(route.id, route);
      }
    } else if (el.type === OSM_TAG_VALUES.TYPE_NODE && el.tags) {
      const node = el as OSMNode;
      const [x, y] = wgs84ToProjected(node.lat, node.lon, targetCRS);
      const building = createBuilding(node.id, [x, y], node.tags || {});
      if (building) {
        buildings.set(building.id, building);
      }
    } else if (el.type === OSM_TAG_VALUES.TYPE_WAY && el.tags && !el.tags.highway) {
      const way2 = el as OSMWay;
      const geometry = buildGeometry(way2.nodes, osmNodes, targetCRS);
      if (geometry.length < GEOMETRY_VALIDATION.MIN_BUILDING_POLYGON_POINTS) continue;

      // Calculate centroid - already in projected coordinates
      const centroid: ProjectedCoords = [
        geometry.reduce((sum, p) => sum + p[0], 0) / geometry.length,
        geometry.reduce((sum, p) => sum + p[1], 0) / geometry.length,
      ];

      const building = createBuilding(way2.id, centroid, way2.tags || {}, geometry);
      if (building) {
        buildings.set(building.id, building);
      }
    }
  }

  return { nodes, links, transportRoutes, buildings, crs: targetCRS };
}

File: map-data-service/parser.py
from __future__ import annotations
from osm_models import OSMApiResponse, OSMNode, OSMWay, OSMRelation
from constants import (
    LINK_TAG_KEYS,
    BUILDING_TAG_KEYS,
    ROUTE_TAG_KEYS,
    VALID_MEMBER_ROLES,
    VALID_BUILDING_TYPES,
)

from models import (
    Building,
    NetworkResponse,
    TrafficLink,
    TrafficNode,
    TransportRoute,
)


def _bucket_elements(
    response: OSMApiResponse,
) -> tuple[dict[int, OSMNode], list[OSMWay], list[OSMRelation]]:
    nodes_by_id: dict[int, OSMNode] = {}
    ways: list[OSMWay] = []
    relations: list[OSMRelation] = []
    for element in response.elements:
        if isinstance(element, OSMNode):
            nodes_by_id[element.id] = element
        elif isinstance(element, OSMWay):
            ways.append(element)
        elif isinstance(element, OSMRelation):
            relations.append(element)
    return nodes_by_id, ways, relations


def _count_endpoint_connections(highway_ways: list[OSMWay]) -> dict[int, int]:
    counts: dict[int, int] = {}
    for way in highway_ways:
        if len(way.nodes) < 2:
            continue
        for node_id in (way.nodes[0], way.nodes[-1]):
            counts[node_id] = counts.get(node_id, 0) + 1
    return counts


def _resolve_geometry(
    node_ids: list[int], nodes_by_id: dict[int, OSMNode]
) -> list[tuple[float, float]] | None:
    coords: list[tuple[float, float]] = []
    for nid in node_ids:
        node = nodes_by_id.get(nid)
        if node is None:
            return None
        coords.append((node.lon, node.lat))
    return coords


def _register_endpoint_nodes(
    way: OSMWay,
    geometry: list[tuple[float, float]],
    traffic_nodes: dict[int, TrafficNode],
    connection_counts: dict[int, int],
) -> None:
    for nid, coord in ((way.nodes[0], geometry[0]), (way.nodes[-1], geometry[-1])):
        if nid not in traffic_nodes:
            traffic_nodes[nid] = TrafficNode(
                id=f"n{nid}",
                osm_id=nid,
                position=coord,
                connection_count=connection_counts.get(nid, 1),
            )


def _make_traffic_link(way: OSMWay, geometry: list[tuple[float, float]]) -> TrafficLink:
    return TrafficLink(
        id=f"l{way.id}",
        osm_id=way.id,
        from_node=f"n{way.nodes[0]}",
        to_node=f"n{way.nodes[-1]}",
        geometry=geometry,
        tags={k: v for k, v in way.tags.items() if k in LINK_TAG_KEYS},
    )


def _build_traffic_graph(
    highway_ways: list[OSMWay],
    nodes_by_id: dict[int, OSMNode],
    connection_counts: dict[int, int],
) -> tuple[list[TrafficNode], list[TrafficLink]]:
    traffic_nodes: dict[int, TrafficNode] = {}
    traffic_links: list[TrafficLink] = []
    for way in highway_ways:
        if len(way.nodes) < 2:
            continue
        geometry = _resolve_geometry(way.nodes, nodes_by_id)
        if geometry is None or len(geometry) < 2:
            continue
        _register_endpoint_nodes(way, geometry, traffic_nodes, connection_counts)
        traffic_links.append(_make_traffic_link(way, geometry))
    return list(traffic_nodes.values()), traffic_links


def _centroid(geometry: list[tuple[float, float]]) -> tuple[float, float]:
    return (
        sum(p[0] for p in geometry) / len(geometry),
        sum(p[1] for p in geometry) / len(geometry),
    )


def _resolve_building_type(tags: dict[str, str]) -> str | None:
    building_type = tags.get("building")
    if building_type == "yes":
        return tags.get("amenity") or tags.get("shop") or None
    return building_type


def _make_building(
    way: OSMWay, geometry: list[tuple[float, float]]
) -> Building | None:
    b_type = _resolve_building_type(way.tags)
    if b_type not in VALID_BUILDING_TYPES:
        return None

    return Building(
        id=f"b{way.id}",
        osm_id=way.id,
        position=_centroid(geometry),
        geometry=geometry,
        type=b_type,
        tags={k: v for k, v in way.tags.items() if k in BUILDING_TAG_KEYS},
    )


def _build_buildings(
    building_ways: list[OSMWay], nodes_by_id: dict[int, OSMNode]
) -> list[Building]:
    buildings: list[Building] = []
    for way in building_ways:
        if len(way.nodes) < 3:
            continue
        geometry = _resolve_geometry(way.nodes, nodes_by_id)
        if geometry is None or len(geometry) < 3:
            continue
        building = _make_building(way, geometry)
        if building:
            buildings.append(building)
    return buildings


def _build_way_geometries(
    ways: list[OSMWay], nodes_by_id: dict[int, OSMNode]
) -> dict[int, list[tuple[float, float]]]:
    geometries: dict[int, list[tuple[float, float]]] = {}
    for way in ways:
        geom = _resolve_geometry(way.nodes, nodes_by_id)
        if geom:
            geometries[way.id] = geom
    return geometries


def _extract_route_members(
    relation: OSMRelation,
    way_geometries: dict[int, list[tuple[float, float]]],
) -> list[TransportRoute]:
    route_tags = {k: v for k, v in relation.tags.items() if k in ROUTE_TAG_KEYS}
    routes: list[TransportRoute] = []
    for member in relation.members:
        if member.type != "way" or member.role not in VALID_MEMBER_ROLES:
            continue
        geom = way_geometries.get(member.ref)
        if geom is not None:
            routes.append(
                TransportRoute(
                    id=f"r{relation.id}_{member.ref}",
                    osm_id=relation.id,
                    way_id=member.ref,
                    geometry=geom,
                    tags=route_tags,
                )
            )
    return routes


def _build_transport_routes(
    relations: list[OSMRelation],
    way_geometries: dict[int, list[tuple[float, float]]],
) -> list[TransportRoute]:
    routes: list[TransportRoute] = []
    for relation in relations:
        if "route" not in relation.tags:
            continue
        routes.extend(_extract_route_members(relation, way_geometries))
    return routes


def parse_osm_response(data: dict) -> NetworkResponse:
    nodes_by_id, ways, relations = _bucket_elements(OSMApiResponse.model_validate(data))

    highway_ways = [w for w in ways if "highway" in w.tags]

    building_ways = [w for w in ways if "building" in w.tags]

    connection_counts = _count_endpoint_connections(highway_ways)

    nodes, links = _build_traffic_graph(highway_ways, nodes_by_id, connection_counts)

    buildings = _build_buildings(building_ways, nodes_by_id)

    way_geometries = _build_way_geometries(ways, nodes_by_id)

    transport_routes = _build_transport_routes(relations, way_geometries)

    return NetworkResponse(
        nodes=nodes,
        links=links,
        buildings=buildings,
        transport_routes=transport_routes,
    )
